components/zcash_address/src/kind/unified.rs

```
@@ -69,51 +100,214 @@ impl Typecode {
+/// an unrecognized metadata item with a typecode in the `MUST-understand` range will result in a
+/// parse failure, instead of the construction of a metadata item.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub enum MetadataItem {
+    /// The expiry height for a Unified Address or Unified Viewing Key
+    ExpiryHeight(u32),
+    /// The expiry time for a Unified Address or Unified Viewing Key
+    ExpiryTime(u64),
+    /// A Metadata Item with an unrecognized Typecode. MUST-understand metadata items are NOT
+    /// represented using this type, as the presence of an unrecognized metadata item with a
+    /// typecode in the `MUST-understand` range will result in a parse failure.
+    Unknown { typecode: u32, data: Vec<u8> },
+}
+
+impl MetadataItem {
+    /// Parse a metadata item for the specified metadata typecode from the provided bytes.
+    pub fn parse(
+        revision: Revision,
+        typecode: MetadataTypecode,
+        data: &[u8],
+    ) -> Result<Self, ParseError> {
+        match (revision, typecode) {
+            (Revision::R1, MetadataTypecode::ExpiryHeight) => data
+                .try_into()
+                .map(u32::from_le_bytes)
+                .map(MetadataItem::ExpiryHeight)
+                .map_err(|_| {
+                    ParseError::InvalidEncoding(
+                        "Expiry height must be a 32-bit little-endian value.".to_string(),
+                    )
+                }),
+            (Revision::R1, MetadataTypecode::ExpiryTime) => data
+                .try_into()
+                .map(u64::from_le_bytes)
+                .map(MetadataItem::ExpiryTime)
+                .map_err(|_| {
+                    ParseError::InvalidEncoding(
+                        "Expiry time must be a 64-bit little-endian value.".to_string(),
+                    )
+                }),
+            (Revision::R1, MetadataTypecode::Unknown(typecode)) => Ok(MetadataItem::Unknown {
```

```suggestion
            (_, MetadataTypecode::Unknown(typecode)) => Ok(MetadataItem::Unknown {
```

---

I have a slight preference to keep the match expressions explicit, because that's more resilient to future changes to the set of revisions (it will cause a compilation failure, rather than silently add default behavior for the new revision.)